maintenance/health/servicehealthcheck/background_healthchecker.go

```diff
@@ -0,0 +1,89 @@
+package servicehealthcheck
+
+import (
+	"context"
+	"time"
+
+	"github.com/opentracing/opentracing-go"
+
+	"github.com/pace/bricks/maintenance/log"
+)
+
+// BackgroundHealthCheck is an interface that, when implemented, will run the
+// HealthCheck in the background, and cache the result, that will be fetched by the HealthHandler
+type BackgroundHealthCheck interface {
+	HealthCheck
+	Interval() time.Duration
+}
+
+// BackgroundHealthCheckMaxWait allows to use a custom amount of time to wait before failing the health check
+type BackgroundHealthCheckMaxWait interface {
+	MaxWait() time.Duration
+}
+
+// registerBackgroundHealthCheck will run the HealthCheck in the background at every given interval.
+// The returned backgroundStateHealthChecker is then used to query this cached state when needed.
+func registerBackgroundHealthCheck(name string, bhc BackgroundHealthCheck) *backgroundStateHealthChecker {
+	var bgState ConnectionState
+
+	go func() {
+		s := log.Sink{Silent: true}
+		ctx := log.ContextWithSink(context.Background(), &s)
+
+		var initErr error
+		if initHC, ok := bhc.(Initializable); ok {
+			if initErr = initHC.Init(ctx); initErr != nil {
+				log.Warnf("error initializing background health check %q: %s", name, initErr)
+				bgState.SetErrorState(initErr)
+			}
+		}
+		timer := time.NewTimer(0) // Do first health check instantly
+		for {
+			func() {
+				<-timer.C
+				defer timer.Reset(bhc.Interval())
+
+				maxWait := cfg.HealthCheckMaxWait
+				if mw, ok := bhc.(BackgroundHealthCheckMaxWait); ok {
```

Comment:
The current implementation of this packages is based on the the following interface:
```go
type HealthCheck interface {
	HealthCheck(ctx context.Context) HealthCheckResult
}
```

At the moment, to add **optional**, but also backward compatible, behavior, the implementation has to implement additional interfaces, eg: 
```go
type Initializable interface {
	Init(ctx context.Context) error
}
```
which allows to have custom initialisation if needed by a specific HealthCheck.

The `BackgroundHealthCheck` follows the same principle, and it contains only the method `Interval() time.Duration` because this is the only one required to make it run in the background. 

The maxWait is optional because the package itself already has a generic maxWait config for all HealthChecks, and to not force every background HC to provide a maxWait method, it was done via another interface that can be implemented as/when needed.

Reply:
Another point to why this might be a better solution than a config variable passed at the beginning, is that having a method, that is called each time, allows the implementation to provide dynamic `Interval` but also `MaxWait` depending on it's current state.

Reply:
Ok.